src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,368 @@
+#define UMThunkStubAMD64_XMM_SAVE_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE - UMThunkStubAMD64_FRAME_OFFSET - UMThunkStubAMD64_XMM_SAVE_NEGOFFSET)
+#define UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE + 8 - UMThunkStubAMD64_FRAME_OFFSET)
+#define UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE (UMThunkStubAMD64_STACK_FRAME_SIZE - (UMThunkStubAMD64_NUM_REG_PUSHES * 8))
+
+#if (UMTHUNKSTUB_HOST_NOTIFY_FLAG_RBPOFFSET - UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET) != 0
+#error update UMTHUNKSTUB_HOST_NOTIFY_FLAG_RBPOFFSET
+#endif
+
+        push_nonvol_reg r12
+        push_nonvol_reg rbp                                                                     // stack_args
+        alloc_stack     UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE
+        set_frame       rbp, UMThunkStubAMD64_FRAME_OFFSET                                      // stack_args
+        mov             byte ptr [rbp + UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET], 0            // hosted
+
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  0h], rdi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET +  8h], rsi
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 10h], rdx
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 18h], rcx
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 20h], r8
+        mov             [rbp + UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET + 28h], r9
+
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET +  0h], xmm0
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 10h], xmm1
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 20h], xmm2
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 30h], xmm3
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 40h], xmm4
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 50h], xmm5
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 60h], xmm6
+        movdqa          [rbp + UMThunkStubAMD64_XMM_SAVE_OFFSET + 70h], xmm7
+
+        mov             [rbp + UMThunkStubAMD64_METHODDESC_SPILL_OFFSET], METHODDESC_REGISTER
+
+        END_PROLOGUE
+
+        //
+        // Call GetThread()
+        //
+        call            C_FUNC(GetThread)
+        mov             METHODDESC_REGISTER, [rbp + UMThunkStubAMD64_METHODDESC_SPILL_OFFSET]
+        test            rax, rax
+        jz              LOCAL_LABEL(DoThreadSetup)
+
+LOCAL_LABEL(HaveThread):
+
+        mov             r12, rax                // r12 <- Thread*
+
+        //
+        // disable preemptive GC
+        //
+        mov             dword ptr [r12 + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 1
+
+        //
+        // catch returning thread here if a GC is in progress
+        //
+        cmp             [rip + C_FUNC(g_TrapReturningThreads)@GOTPCREL], 0
```

Comment:
PREPARE_EXTERNAL_VAR loads the address of the var into a helper reg - I think a straight cmp will compare address of the var with 0.